File: backend/services/llm_service.py
import os
import json
import logging
from typing import List, Dict, Any

logger = logging.getLogger(__name__)

class LLMService:
    """
    Mock LLM Service - Ready for LangGraph Agent Integration
    
    This service provides realistic mock responses that match your frontend
    expectations. When ready, replace with your LangGraph AgentChatService.
    
    TODO: Replace with your LangGraph AgentChatService:
    from your_agent_module import AgentChatService
    """
    
    def __init__(self):
        # TODO: Replace with your LangGraph AgentChatService
        # self.agent = AgentChatService()
        pass
    
    def analyze_query(self, query: str, message_history: List[Dict] = None) -> str:
        """
        Analyze user query for intent and criteria extraction.
        
        Args:
            query: User's query string
            message_history: Previous conversation messages
            
        Returns:
            JSON string with analysis results
        """
        logger.info(f"LLMService: Analyzing query: {query}")
        
        # Mock analysis based on query content
        query_lower = query.lower()
        
        # Determine if property-related
        property_keywords = ["property", "house", "flat", "apartment", "home", "property", "comp", "comparable", "valuation", "price", "bedroom", "bathroom"]
        is_property_related = any(keyword in query_lower for keyword in property_keywords)
        
        # Extract basic criteria (mock extraction)
        extracted_criteria = {}
        
        # Look for bedroom count
        import re
        bedroom_match = re.search(r'(\d+)\s*bed(room)?', query_lower)
        if bedroom_match:
            extracted_criteria["bedrooms"] = int(bedroom_match.group(1))
        
        # Look for bathroom count  
        bathroom_match = re.search(r'(\d+)\s*bath(room)?', query_lower)
        if bathroom_match:
            extracted_criteria["bathrooms"] = int(bathroom_match.group(1))
        
        # Look for price range
        price_match = re.search(r'£?(\d+)(?:k|,000)?', query_lower)
        if price_match:
            price = int(price_match.group(1))
            if 'k' in query_lower or ',000' in query_lower:
                price *= 1000
            extracted_criteria["max_price"] = price
        
        # Look for location
        location_keywords = ["london", "bristol", "manchester", "birmingham", "leeds", "glasgow", "edinburgh"]
        for location in location_keywords:
            if location in query_lower:
                extracted_criteria["location"] = location.title()
                break
        
        # Determine response type
        if is_property_related:
            if "comp" in query_lower or "comparable" in query_lower:
                response_type = "comparable_search"
                suggested_response = "I'll help you find comparable properties. Let me search our database for similar properties in your area."
            else:
                response_type = "property_search"
                suggested_response = "I'll help you with property information. Let me search for properties that match your criteria."
        else:
            response_type = "general_query"
            suggested_response = "I'm here to help with property-related questions. Could you tell me more about what you're looking for?"
        
        # Mock response data
        response_data = {
            "intent": "property_search" if is_property_related else "general_question",
            "extracted_criteria": extracted_criteria,
            "confidence": 0.85,
            "response_type": response_type,
            "suggested_response": suggested_response,
            "needs_clarification": len(extracted_criteria) == 0 and is_property_related,
            "missing_information": [] if extracted_criteria else ["location", "property_type"] if is_property_related else []
        }
        
        logger.info(f"LLMService: Analysis complete - {response_type}")
        return json.dumps(response_data)
    
    def chat_completion(self, messages: List[Dict[str, str]]) -> Dict[str, Any]:
        """
        Generate AI chat response.
        
        Args:
            messages: List of message objects with 'role' and 'content'
            
        Returns:
            Dictionary with response data
        """
        logger.info(f"LLMService: Chat completion with {len(messages)} messages")
        
        if not messages:
            return {
                "message": "Hello! I'm your property assistant. How can I help you today?",
                "role": "assistant"
            }
        
        # Get the last user message
        last_message = None
        for msg in reversed(messages):
            if msg.get('role') == 'user':
                last_message = msg.get('content', '')
                break
        
        if not last_message:
            last_message = "Hello"
        
        # Mock response based on message content
        message_lower = last_message.lower()
        
        if any(word in message_lower for word in ["hello", "hi", "hey", "good morning"]):
            response = "Hello! I'm your property assistant. I can help you find properties, analyze comparables, and answer questions about real estate. What would you like to know?"
        elif any(word in message_lower for word in ["thank", "thanks"]):
            response = "You're welcome! Is there anything else I can help you with regarding properties?"
        elif "property" in message_lower or "house" in message_lower:
            response = "I'd be happy to help you with property information! Could you tell me more about what you're looking for? For example, are you interested in finding comparables, getting a valuation, or searching for specific types of properties?"
        else:
            response = f"I understand you're asking about '{last_message}'. As a property assistant, I'm here to help with real estate questions, property searches, and market analysis. Could you provide more details about what you'd like to know?"
        
        return {
            "message": response,
            "role": "assistant",
            "timestamp": "2024-01-01T12:00:00Z"  # Mock timestamp
        }

File: frontend-ts/src/llm/llmService.ts
/**
 * LLM Service (Secure Backend Proxy)
 * All LLM requests now route through Flask backend to keep API keys secure
 * This supports your multi-agent backend system
 */

import { backendApi } from '../services/backendApi';

export interface LLMAnalysisResult {
  isPropertyRelated: boolean;
  needsClarification: boolean;
  extractedCriteria: {
    bedrooms?: number;
    bathrooms?: number;
    location?: string;
    priceRange?: { min?: number; max?: number };
  };
  responseType: 'property_search' | 'general_response' | 'clarification' | 'content_creation' | 'data_analysis';
  suggestedResponse?: string;
}

export interface Message {
  id: string;
  content: string;
  role: 'user' | 'assistant';
  timestamp: Date;
}

// LLM Configuration (now managed by backend)
export const LLM_CONFIG = {
  provider: 'backend', // All LLM calls go through Flask backend
  note: 'LLM configuration and API keys are managed securely by the backend service',
};

/**
 * Main LLM Analysis Function
 * Routes through backend which handles multi-agent system
 */
export const analyzeQueryWithLLM = async (
  query: string,
  messageHistory: Message[]
): Promise<LLMAnalysisResult> => {
  try {
    const response = await backendApi.analyzeQuery(query, messageHistory);

    if (!response.success || !response.data) {
      console.warn('Backend LLM analysis failed, using fallback');
      return await mockLLMAnalysis(query, messageHistory);
    }

    return response.data as LLMAnalysisResult;
  } catch (error) {
    console.error('LLM analysis error:', error);
    return await mockLLMAnalysis(query, messageHistory);
  }
};

/**
 * Fallback Mock Analysis
 * Used when backend is unavailable or for development
 */
const mockLLMAnalysis = async (query: string, messageHistory: Message[]): Promise<LLMAnalysisResult> => {
  const lowerQuery = query.toLowerCase();
  
  // Simple property detection
  const propertyKeywords = [
    'property', 'properties', 'comp', 'comps', 'comparable', 'comparables', 
    'house', 'houses', 'home', 'homes', 'real estate', 'listing', 'listings', 
    'bed', 'bedroom', 'bedrooms', 'bath', 'bathroom', 'bathrooms'
  ];
  const isPropertyRelated = propertyKeywords.some(keyword => lowerQuery.includes(keyword));

  if (!isPropertyRelated) {
    // Check if it's a content creation request
    const contentKeywords = ['write', 'create', 'help me write', 'draft', 'compose', 'generate', 'refine', 'edit', 'improve', 'rewrite'];
    const isContentRequest = contentKeywords.some(keyword => lowerQuery.includes(keyword));
    
    if (isContentRequest) {
      return {
        isPropertyRelated: false,
        needsClarification: false,
        extractedCriteria: {},
        responseType: 'content_creation',
        suggestedResponse: "I can help you write property descriptions. What property details do you have?"
      };
    }
    
    return {
      isPropertyRelated: false,
      needsClarification: false,
      extractedCriteria: {},
      responseType: 'general_response',
      suggestedResponse: "Hello. How can I help you today?"
    };
  }

  // Extract criteria
  const bedroomMatch = lowerQuery.match(/(\d+)\s*(?:bed|bedroom|bedrooms)/);
  const bathroomMatch = lowerQuery.match(/(\d+)\s*(?:bath|bathroom|bathrooms)/);
  const locationMatch = lowerQuery.match(/(?:in|at|near|around)\s+([a-zA-Z\s]+?)(?:\s|$|,|\.)/);
  
  const extractedCriteria = {
    bedrooms: bedroomMatch ? parseInt(bedroomMatch[1]) : undefined,
    bathrooms: bathroomMatch ? parseInt(bathroomMatch[1]) : undefined,
    location: locationMatch ? locationMatch[1].trim().toLowerCase() : undefined,
    priceRange: undefined
  };

  // Check if needs clarification
  const vagueTerms = ['comps', 'comp', 'comparable', 'comparables', 'properties', 'property'];
  const hasVagueTerms = vagueTerms.some(term => lowerQuery.includes(term));
  const hasSpecificTerms = bedroomMatch || bathroomMatch || locationMatch || 
                          lowerQuery.includes('price') || lowerQuery.includes('value');
  
  const needsClarification = hasVagueTerms && !hasSpecificTerms;

  if (needsClarification) {
    return {
      isPropertyRelated: true,
      needsClarification: true,
      extractedCriteria: {},
      responseType: 'clarification',
      suggestedResponse: "I can help you find property comparables. Please specify: bedrooms, location, and price range."
    };
  }

  return {
    isPropertyRelated: true,
    needsClarification: false,
    extractedCriteria,
    responseType: 'property_search',
    suggestedResponse: `Found properties matching your criteria.`
  };
};
